vault/resource_plugin.go

```diff
@@ -0,0 +1,196 @@
+				Type:        schema.TypeString,
+				Description: "Semantic version of the plugin.",
+				Optional:    true,
+				ForceNew:    true,
+			},
+			fieldSHA256: {
+				Type:        schema.TypeString,
+				Description: "The SHA256 sum of the plugin binary.",
+				Required:    true,
+				ForceNew:    true,
+			},
+			fieldCommand: {
+				Type:        schema.TypeString,
+				Description: "The command to execute the plugin, relative to the plugin_directory.",
+				Required:    true,
+				ForceNew:    true,
+			},
+			fieldArgs: {
+				Type:        schema.TypeList,
+				Description: "Any additional arguments to pass to the plugin.",
+				Optional:    true,
+				ForceNew:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			fieldEnv: {
+				Type:        schema.TypeList,
+				Description: "Any additional environment variables to run the plugin with.",
+				Optional:    true,
+				ForceNew:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			fieldOCIImage: {
+				Type:        schema.TypeString,
+				Description: "Specifies OCI image to run. If specified, setting command, args, and env will update the container's entrypoint, args, and environment variables (append-only) respectively.",
+				Optional:    true,
+				ForceNew:    true,
+			},
+			fieldRuntime: {
+				Type:        schema.TypeString,
+				Description: "Specifies Vault plugin runtime to use if oci_image is specified.",
+				Optional:    true,
+				ForceNew:    true,
+			},
+		},
+	}
+}
+
+func pluginWrite(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := provider.GetClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	pluginType, err := api.ParsePluginType(d.Get(consts.FieldType).(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	name := d.Get(consts.FieldName).(string)
+	version := d.Get(consts.FieldVersion).(string)
+	id := fmt.Sprintf("%s/%s", pluginType, name)
+	if version != "" {
+		id = fmt.Sprintf("%s/%s", id, version)
+	}
+
+	log.Printf("[DEBUG] Writing plugin %q", id)
+	err = client.Sys().RegisterPluginWithContext(ctx, &api.RegisterPluginInput{
+		Type:     pluginType,
+		Name:     name,
+		Version:  version,
+		SHA256:   d.Get(fieldSHA256).(string),
+		Command:  d.Get(fieldCommand).(string),
+		Args:     util.ToStringArray(d.Get(fieldArgs).([]interface{})),
+		Env:      util.ToStringArray(d.Get(fieldEnv).([]interface{})),
+		OCIImage: d.Get(fieldOCIImage).(string),
+		Runtime:  d.Get(fieldRuntime).(string),
+	})
+	if err != nil {
+		return diag.Errorf("error updating plugin %q: %s", id, err)
+	}
+	log.Printf("[DEBUG] Wrote plugin %q", id)
+
+	d.SetId(id)
+
+	return nil
+}
+
+func pluginRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	pluginType, err := api.ParsePluginType(d.Get(consts.FieldType).(string))
```

Comment:
Thanks for adding feature support for these!

Sorry, I know this is still in draft, but I was interested in taking a look and noticed the use of `d.Get` within the read function. This will cause the import step to fail, since `d` is empty during that step. Since the `type`, `name` and `version` are encoded in the ID, we can extract that info from the ID using a regex

Reply:
Thanks for taking a look! That kind of feedback is exactly why I opened it early rather than waiting until it was finished :) if you do take another pass please do feel free to leave comments on anything else that isn't best practice etc, this is the first work I've done in TFVP for a while so I feel a bit out of date.